pennylane/compiler/python_compiler/transforms/mbqc/split_non_commuting.py

```diff
@@ -0,0 +1,396 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This file contains the implementation of the split_non_commuting transform."""
+
+from dataclasses import dataclass
+from typing import Type, TypeVar
+
+from xdsl import context, passes, pattern_rewriter
+from xdsl.dialects import builtin, func
+from xdsl.ir import Operation, SSAValue
+from xdsl.rewriter import InsertPoint
+
+from pennylane.compiler.python_compiler import compiler_transform
+from pennylane.compiler.python_compiler.dialects import quantum
+
+
+@dataclass(frozen=True)
+class SplitNonCommutingPass(passes.ModulePass):
+    """Pass that splits non-commuting regions in a quantum function."""
+
+    name = "split-non-commuting"
+
+    def apply(self, _ctx: context.Context, op: builtin.ModuleOp) -> None:
+        self.apply_on_qnode(op, SplitNonCommutingPattern())
+
+    def apply_on_qnode(self, module: builtin.ModuleOp, pattern: pattern_rewriter.RewritePattern):
+        """Apply given pattern once to the QNode function in this module."""
+        rewriter = pattern_rewriter.PatternRewriter(module)
+        for op in module.ops:
+            if isinstance(op, func.FuncOp) and "qnode" in op.attributes:
+                pattern.match_and_rewrite(op, rewriter)
+
+
+split_non_commuting_pass = compiler_transform(SplitNonCommutingPass)
+
+
+class SplitNonCommutingPattern(pattern_rewriter.RewritePattern):
+    """Split non-commuting region into multiple regions"""
+
+    def __init__(self):
+        self.module: builtin.ModuleOp = None
+
+    T = TypeVar("T")
+
+    def get_parent_of_type(self, op: Operation, kind: Type[T]) -> T | None:
+        """Walk up the parent tree until an op of the specified type is found."""
+        while (op := op.parent_op()) and not isinstance(op, kind):
+            pass
```

Comment:
So, is it correctly understood that if we don't find an `op` of the specified type, we get to the very top parent op, and it has `op.parent_op() is None`, so the `while` loop exits and the very top operation is returned?